src/hardware/agilent-54621d/protocol.c

```diff
@@ -0,0 +1,1032 @@
+	{ 200, 1000 },
+	{ 500, 1000 },
+	/* volts */
+	{ 1, 1 },
+	{ 2, 1 },
+	{ 5, 1 },
+	{ 10, 1 },
+	{ 20, 1 },
+	{ 50, 1 },
+	{ 100, 1 },
+};
+
+static const char *scope_analog_channel_names[] = {
+	"CHAN1", "CHAN2",
+};
+
+static const char *scope_digital_channel_names[] = {
+	"D0", "D1", "D2", "D3", "D4", "D5", "D6", "D7",
+	"D8", "D9", "D10", "D11", "D12", "D13", "D14", "D15",	
+}; 
+
+static struct scope_config scope_models[] = {
+	{
+		/* Agilent 54621D/54622D Models only differ in Bandwidth; everything else should be the same*/
+		.name = {"54621D", "54622D", NULL},
+		.analog_channels = 2,
+		.digital_channels = 16,
+
+		.analog_names = &scope_analog_channel_names,
+		.digital_names = &scope_digital_channel_names,
+
+		.devopts = &devopts,
+		.num_devopts = ARRAY_SIZE(devopts),
+
+		.devopts_cg_analog = &devopts_cg_analog,
+		.num_devopts_cg_analog = ARRAY_SIZE(devopts_cg_analog),
+
+		.devopts_cg_digital = &devopts_cg_digital,
+		.num_devopts_cg_digital = ARRAY_SIZE(devopts_cg_digital),
+
+		.digital_pods = 2,
+
+		.coupling_options = &coupling_options,
+		.num_coupling_options = ARRAY_SIZE(coupling_options),
+
+		.logic_threshold = &logic_threshold,
+		.num_logic_threshold = ARRAY_SIZE(logic_threshold),
+		.logic_threshold_for_pod = TRUE,
+
+		.trigger_sources = &trigger_sources,
+		.num_trigger_sources = ARRAY_SIZE(trigger_sources),
+
+		.trigger_slopes = &scope_trigger_slopes,
+		.num_trigger_slopes = ARRAY_SIZE(scope_trigger_slopes),
+
+		.timebases = &scope_timebases,
+		.num_timebases = ARRAY_SIZE(scope_timebases),
+
+		.vdivs = &vdivs,
+		.num_vdivs = ARRAY_SIZE(vdivs),
+
+		.num_xdivs = 10,
+		.num_ydivs = 8,
+
+		.scpi_dialect = &agilent_scpi_dialect,
+	},
+	//ToDo: Implement other scope models
+};
+
+SR_PRIV int agilent_54621d_update_sample_rate(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct scope_state *state;
+	const struct scope_config *config;
+	int tmp_int;
+
+	devc = sdi->priv;
+	config = devc->model_config;
+	state = devc->model_state;
+
+	if (sr_scpi_get_int(sdi->conn,
+			      (*config->scpi_dialect)[SCPI_CMD_GET_SAMPLE_RATE],
+			      &tmp_int) != SR_OK)
+		return SR_ERR;
+
+	state->sample_rate = tmp_int;
+
+	if(devc->sample_rate_limit > (uint64_t)tmp_int)
+		devc->sample_rate_limit = tmp_int;
+
+	return SR_OK;
+}
+
+SR_PRIV int agilent_54621d_init_device(struct sr_dev_inst *sdi)
+{
+	int model_index;
+	unsigned int i, j, group;
+	struct sr_channel *ch;
+	struct dev_context *devc;
+	const char *cg_name;
+	int ret;
+
+	devc = sdi->priv;
+	model_index = -1;
+
+	/* find model  */
+	for(i = 0; i < ARRAY_SIZE(scope_models); i++) {
+		for(j = 0; scope_models[i].name[j]; j++) {
+			if(!strcmp(sdi->model, scope_models[i].name[j])) {
+				model_index = i;
+			}
+		}
+		if(model_index != -1)
+			break;
+	}
+
+	if(model_index == -1){
+		sr_dbg("Unsupported device.");
+		return SR_ERR_NA;
+	}
+
+	devc->analog_groups = g_malloc0(sizeof(struct sr_channel_group*) *scope_models[model_index].analog_channels);
```

Comment:
current project guidelines say

> We assume that "small" memory allocations (< 1MB) will always succeed.
   Thus, it's fine to use g_malloc() or g_malloc0() for allocations of
   simple/small structs and such (instead of using g_try_malloc()), and
   there's no need to check the return value.